### DISCUSSION TITLE
Tree-sitter integration

### DISCUSSION THREAD

**Discussion by bmulholland** [ggandor/lightspeed.nvim#92] [Ideas]
I'm taking a look at https://github.com/mfussenegger/nvim-treehopper and wondering if this is something that lightspeed is thinking about incorporating?

**Comment by ggandor** [comment]
The big question is, (how) can we utilize Tree-sitter in an organic way? We could e.g. reimplement Treehopper and bundle it to Lightspeed, but then all we've done is reimplementing Treehopper and bundling it to Lightspeed... I mean, it's an orthogonal feature (the highlighting style can be unified, but that's all).

I've already been pondering about this, and even had some ideas, but neither of them turned around to be viable or particularly useful so far. On this note, crazy ideas from anyone are more than welcome here. ;)

**Reply by ggandor** [reply]
An extension plugin to select tree-sitter nodes would be no biggie though (**photosynthesis.nvim**? 😀)

**Reply by bmulholland** [reply]
Harmonization of UI/UX would be one thing I'd be looking for here. Highlighting style, key selection, but probably also the handling of repeated items. Perhaps that could be done as a plugin or other type of integration to avoid duplicating work, as you say.

However: there may be something much more interesting at a higher level. Depending on how interested you are in taking this on (I certainly don't have time, and perhaps this stuff /does/ belong in treehopper 🙈), some ideas that come to mind:
1. Once a variable is selected using `s`, then a second-level selection comes up to get to the repetition/use of that same var.
2. First choose whether `s` should look for variables, functions, constants, comments, or something else. If these are always the same key, then, say, `sa" to select a variable could be quite instinctive and fast (once learned). I usually know which of these I'm looking for, so narrowing the search space /may/ become faster and more intuitive.
3. A framework that enables plugins to add functionality like these, because others may be able to come up with better ideas once the details are abstracted away.

(Perhaps some or all of this should be on a different key than `s`?)

IMO, an extension plugin seems like a good balance to try some experiments that can be (more easily) discarded if it turns out to be not too useful/too complex/superseded by something else.
